[Feat] 애플리케이션 배포 환경 구축을 위한 Docker Compose 추가하기

langoustinee opened the pull request:
## 🤔 Motivation
- 개발환경 배포 전 이미지 정상 여부 테스트를 위한 docker-compose 확장

<br>

## 💡 Key Changes

### dockerfile
- OpenJDK로 사용할 베이스 이미지의 경우 `openjdk:17-jdk-slim`와 같은 일반적인 이미지가 아닌 `bellsoft/liberica-openjdk-alpine:17` 이미지를 사용하였습니다. `bellsoft/liberica-openjdk-alpine:17` 이미지는 Alpine Linux를 기반으로 하여, 이미지 크기를 최소화하는 것에 초점을 맞춰 경량화되었기에 테스트 단계에서 보다 빠르고 가볍게 이미지화하는데 도움이 될 것으로 판단하여 해당 이미지를 사용했습니다.

### docker-compose
- Spring Application과 MariaDB를 함께 구동하기 위해 Docker Compose로 설정을 해두었습니다. 
- 애플리케이션은 8080 포트로, MariaDB는 3306포트로 구동됩니다.
- 추후 DB 볼륨 설정이 필요할 경우 volumes 설정을 확장하면 됩니다.

### 로컬 환경 실행 가이드
1. 도커가 로컬 환경에 구비되어 있어야 합니다. 본 프로젝트에서는 Docker Desktop을 이용하기로 결정했기에 Docker Desktop을 설치하시면 됩니다.
2. 프로젝트 최상위 경로에 .env 파일을 만들어 환경변수를 작성합니다. (자세한 환경변수 내용은 노션에 최신화 해두었습니다.)
3. 프로젝트 최상위 경로에서 애플리케이션을 빌드하여 jar 파일을 준비합니다. (_빌드 성공하면 build/lib 디렉토리 하위에 jar파일이 생성되었는지 확인합니다._)
4. 프로젝트 최상위 경로에서 아래 명령을 수행합니다.

```bash
# 첫번째 명령
./gradlew clean build

# 두번째 명령
docker compose up -d
```

5. 약 30초 정도 배포 시간 소요 후, docker ps 명령이나, Docker Desktop UI로 애플리케이션 컨테이너와 MariaDB 컨테이너가 정상적으로 실행되고 있는지 확인합니다.

<img width="1250" alt="image" src="https://github.com/four-uncles/saramara-community-server/assets/59594946/e1c24ec5-aa39-4895-bcd7-fc73abe298ec">

<br>

## 👨‍👨‍👧‍👦 To Reviewers
- @byeongJoo05 @IToriginal 

close #157

langoustinee committed:
✨feat: 애플리케이션 Dockerfile 추가하기

langoustinee committed:
✨feat: 단위 이미지 패키지 제공을 위한 Docker Compose 확장하기

langoustinee committed:
🔧fix: 개발환경 배포 테스트를 위한 프로파일 변경하기

- local 환경에서 dev 환경으로 설정 변경

langoustinee committed:
🔧fix: 내부 env 파일 ignore 추가하기

byeongJoo05 commented on src/main/resources/application.yml:
```diff
@@ -1,11 +1,11 @@
 spring:
   profiles:
-    active: local
+    active: dev
```
프로젝트 내 profiles active는 `local`로 두고, compose yml 내부에 environment 설정에 `SPRING_PROFILES_ACTIVE: DEV` 로 두는 것은 어떤가요?

아직 개발 서버와 로컬 서버 내 차이는 없지만 위 설명과 같이 운영환경과 프로젝트 내 설정을 따로할 수 있다는 장점이 있습니다!

langoustinee replied:
엇, application.yml 파일에 말씀해주신대로 프로파일은 local 그대로 두고, 도커 빌드시 dev 프로파일로 설정하도록 개발했었는데, 트러블 슈팅하다가 롤백되어버렸나 봅니다 ㅠㅠ 

해당 설정 프로파일은 local로 돌려두고, 빌드시 dev 프로파일로 빌드되도록 다시 수정해두겠습니다. 날카로운 피드백 감사해유!!

byeongJoo05 commented:
사적 질문이지만 가끔가다 특정 Mac OS 문제인지 감은 안잡히지만 arm64 기준으로 도커파일을 실행하였을 때, Linux 서버 내에서 컨테이너화 도중 java platform을 못 찾는 문제가 간간히 일어난다고 하더라구요... 혹시 관련된 원인을 아시나요?

`--platform linux/amd64`과 같은 명령어를 통해 build 주체가 누구든 같은 플랫폼기반으로 패키징이 가능한 듯 보이는데 어떻게 생각하시는 지 궁금합니다.

IToriginal reviewed (dismissed):
수고 많으셨습니다.

주석을 활용해 각 라인의 기능을 작성해주셔서 각 라인에 코드들이 어떤 기능을 뜻하는지 이해하기 편했습니다!

감사합니다!

langoustinee commented:
@byeongJoo05 이미지 OS 플랫폼 관련 질문 저도 최근에 겪었던 이슈라 아는데로 공유드려봅니다~ 큐큐

> 사적 질문이지만 가끔가다 특정 Mac OS 문제인지 감은 안잡히지만 arm64 기준으로 도커파일을 실행하였을 때, Linux 서버 내에서 컨테이너화 도중 java platform을 못 찾는 문제가 간간히 일어난다고 하더라구요... 혹시 관련된 원인을 아시나요?
--platform linux/amd64과 같은 명령어를 통해 build 주체가 누구든 같은 플랫폼기반으로 패키징이 가능한 듯 보이는데 어떻게 생각하시는 지 궁금합니다.

일단 **Mac OS M1은 기본적으로 arm 기반 아키텍처**입니다. 그런데 보통 일반적으로 사용되는 AWS EC2와 같은 VM 인스턴스는 대부분 amd 기반의 아키텍처를 차용하기 때문에 Mac OS M1에서 빌드한 이미지를 범용적으로 사용하려면 docker build 명령에 `--platform linux/amd64` 옵션을 함께 부여하여 빌드해야 하더라구요.

말씀해주신 김에 저희 docker-compose 설정에도 docker build시 linux/amd64 옵션을 부여하여 빌드하도록 수정해두겠습니다. 감사합니다 :)

langoustinee commented:
@byeongJoo05 @IToriginal 프로파일 설정 수정과 linux/amd64 플랫폼 옵션 부여를 위해 approve를 다시 요청드렸습니다! 
해당 이슈개발 후 다시 커밋해두겠으니 확인 부탁드려요~

langoustinee committed:
🔧 fix: application.yml 프로파일 기본값으로 원복하기

- 로컬 개발환경을 기본값으로 가져가기 위함.

langoustinee committed:
🔧 fix: docker-compose 설정 변경

- DB Health Check 추가하여 DB 안정화 이후 Spring Application 실행할 수 있도록 반영
- Spring Boot Profile을 dev로 적용
- DB 이미지 버전을 10.11 버전으로 사용하도록 지정

langoustinee committed:
🔧 fix: Dockerfile 내용 임시 수정

- docker-compose로 gradle clean build 명령시 테스트 실패하는 이슈 때문에 로컬에서 먼저 빌드한 jar 파일로 이미지를 빌드할 수 있도록 임시로 수정함.
